storage/postgres/storage.sql

```diff
@@ -0,0 +1,138 @@
+-- Postgres impl of storage
+-- ---------------------------------------------
+-- Tree stuff here
+-- ---------------------------------------------
+
+-- Tree Enums
+CREATE TYPE E_TREE_STATE AS ENUM('ACTIVE', 'FROZEN', 'DRAINING');
+CREATE TYPE E_TREE_TYPE AS ENUM('LOG', 'MAP', 'PREORDERED_LOG');
+CREATE TYPE E_HASH_STRATEGY AS ENUM('RFC6962_SHA256', 'TEST_MAP_HASHER', 'OBJECT_RFC6962_SHA256', 'CONIKS_SHA512_256', 'CONIKS_SHA256');
+CREATE TYPE E_HASH_ALGORITHM AS ENUM('SHA256');
+CREATE TYPE E_SIGNATURE_ALGORITHM AS ENUM('ECDSA', 'RSA');
+
+-- Tree parameters should not be changed after creation. Doing so can
+-- render the data in the tree unusable or inconsistent.
+CREATE TABLE IF NOT EXISTS trees ( 
+  tree_id                  BIGINT NOT NULL, 
+  tree_state               E_TREE_STATE NOT NULL, 
+  tree_type                E_TREE_TYPE NOT NULL, 
+  hash_strategy            E_HASH_STRATEGY NOT NULL, 
+  hash_algorithm           E_HASH_ALGORITHM NOT NULL, 
+  signature_algorithm      E_SIGNATURE_ALGORITHM NOT NULL, 
+  display_name             VARCHAR(20), 
+  description              VARCHAR(200), 
+  create_time_millis       BIGINT NOT NULL, 
+  update_time_millis       BIGINT NOT NULL, 
+  max_root_duration_millis BIGINT NOT NULL, 
+  private_key              BYTEA NOT NULL, 
+  public_key               BYTEA NOT NULL, 
+  deleted                  BOOLEAN, 
+  delete_time_millis       BIGINT, 
+  PRIMARY KEY(tree_id) 
+); 
+
+-- This table contains tree parameters that can be changed at runtime such as for
+-- administrative purposes.
+CREATE TABLE IF NOT EXISTS tree_control(
+  tree_id                   BIGINT NOT NULL,
+  signing_enabled           BOOLEAN NOT NULL,
+  sequencing_enabled        BOOLEAN NOT NULL,
+  sequence_interval_seconds INTEGER NOT NULL,
+  PRIMARY KEY(tree_id),
+  FOREIGN KEY(tree_id) REFERENCES trees(tree_id) ON DELETE CASCADE
+);
+
+CREATE TABLE IF NOT EXISTS subtree(
+  tree_id               BIGINT NOT NULL,
+  subtree_id            BYTEA NOT NULL,
+  nodes                 BYTEA NOT NULL,
+  subtree_revision      INTEGER NOT NULL, 
+  PRIMARY KEY(tree_id, subtree_id, subtree_revision),
+  FOREIGN KEY(tree_id) REFERENCES Trees(tree_id) ON DELETE CASCADE
+);
+
+-- The TreeRevisionIdx is used to enforce that there is only one STH at any
+-- tree revision
+CREATE TABLE IF NOT EXISTS tree_head(
+  tree_id                BIGINT NOT NULL,
+  tree_head_timestamp    BIGINT,
+  tree_size              BIGINT,
+  root_hash              BYTEA NOT NULL,
+  root_signature         BYTEA NOT NULL,
+  tree_revision          BIGINT,
+  PRIMARY KEY(tree_id, tree_head_timestamp),
```

Comment:
I _think_ it would be better to have the PK be `tree_id, tree_revision`, you could then drop the `index` below (I don't think anything needs this key to contain the timestamp - the cloudspanner implementation uses `/treeID/tree_revision` as it's PK on the equivalent table there BTW).

Separately, and I must admit I don't know much about PGSQL, but other storage impls define the PK here as `tree_id, tree_rev DESC`, because by far the most common operation on this table is a scan which  wants to fetch the newest entry.  It looks like PG doesn't support the `DESC` modifier on the PK fields, so I don't know what implications this will have for performance when you come to write the code itself.

Might be worth popping a comment or a `TODO` here as a note to yourself to look at this area again in detail once the code lands too.